### PR TITLE
Frames: Reduce Empty Supercell Memory Foot Print

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- the last frame in the supercells was never removed because of the [short circuit usage](http://en.wikipedia.org/wiki/Short-circuit_evaluation)
  `... = isValid && ...;`
- the effect of this bug is a larger memory foot print of an empty (zero particles) supercell

*This bug effects all previous releases but is not critical!*

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
thx @BenjaminW3 to point me to this.

**Comment by ax3l** [comment]
well spotted you two! :sparkles:

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 1]
```diff
@@ -392,26 +403,27 @@ __global__ void kernelFillGaps(ParticlesBox<FRAME, Mapping::Dim> pb, Mapping map
...
             else if (counterGaps == counterParticles)
             {
-                lastFrame = &(pb.getPreviousFrame(*lastFrame, isValid));
-                isValid = isValid && pb.removeLastFrame(superCellIdx);
-                if (isValid && lastFrame != firstFrame)
+                isFrameValid = getPreviousFrameAndRemoveLastFrame(lastFrame, pb, superCellIdx);
+                if (isFrameValid && lastFrame != firstFrame)
                 {
-                    firstFrame = &(pb.getNextFrame(*firstFrame, isValid));
+                    firstFrame = &(pb.getNextFrame(*firstFrame, isFrameValid));
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: some vals like `firstFrame` are manipulated here but until the next `while` loop there is no `__syncthreads();` called - sure that is still correct? looks like a race condition to me.

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 2]
```diff
@@ -392,26 +403,27 @@ __global__ void kernelFillGaps(ParticlesBox<FRAME, Mapping::Dim> pb, Mapping map
...
                 }
+
             }
+            isValid = isFrameValid;
         }
         __syncthreads();
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: ah here it is -> remove comments

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 3]
```diff
@@ -482,11 +494,6 @@ __global__ void kernelBashParticles(ParticlesBox<FRAME, Mapping::Dim> pb,
 
     DataSpace<Dim> superCellIdx = mapper.getSuperCellIndex(DataSpace<Dim > (blockIdx));
 
-    // asymmetric, because corners are exchanged only in y-directions
-    /*   const DataSpace<Dim> directions = Mask::getRelativeDirections<Dim > (mapper.getExchangeType());
-
-       if (directions.y() != 0 && directions.x() == 0)
-           superCellIdx.x() -= mapper.getGuardingSuperCells();*/
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: why did you remove this lines and how is it related to the bug fix?

Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: ah, dead code.
